### PR TITLE
 Frontend/Render Mode: Feature - Split static render and regular render.

### PR DESCRIPTION
**Pull request by Nevoss**

### CODE REVIEW

File: core/frontend/render-mode-manager.php [hunk 1]
```diff
@@ -75,12 +75,23 @@ public function get_current() {
 		return $this->current;
 	}
 
+	/**
+	 * @param Render_Mode_Base $render_mode
+	 *
+	 * @return $this
+	 */
+	public function set_current( Render_Mode_Base $render_mode ) {
```
Comment (matipojo) [inline core/frontend/render-mode-manager.php:83]:
```suggestion
	private function set_current( Render_Mode_Base $render_mode ) {
```

File: tests/phpunit/elementor/includes/base/test-widget-base.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+		ob_start();
+
+		$widget->render_content();
+
+		$this->assertRegExp('/render/', ob_get_clean());
+	}
+}
+
+class Static_Render_Mode extends Render_Mode_Base {
```
Comment (matipojo) [inline tests/phpunit/elementor/includes/base/test-widget-base.php:47]: Move to a file in the mock folder.

**Review by matipojo — CHANGES_REQUESTED** [review]

File: core/frontend/render-mode-manager.php (hunk 1 above)
Comment (Nevoss) [reply core/frontend/render-mode-manager.php:83]: This one creates a quick way to define the render mode without any $_GET params.

**Review by matipojo — CHANGES_REQUESTED** [review]

**Review by matipojo — DISMISSED** [review]

File: includes/base/widget-base.php [hunk 3]
```diff
@@ -527,7 +527,7 @@ public function render_content() {
 			$skin->set_parent( $this );
 			$skin->render();
 		} else {
-			$this->render();
```
Comment (kobizz) [inline includes/base/widget-base.php:530]: What if a skin requires a static render?

**Review by kobizz — CHANGES_REQUESTED** [review]

File: tests/phpunit/elementor/includes/base/test-widget-base.php [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+use Elementor\Tests\Phpunit\Includes\Base\Mock\Mock_Widget;
+use Elementor\Tests\Phpunit\Includes\Base\Mock\Mock_Static_Render_Mode;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+class Test_Widget_Base extends Elementor_Test_Base {
+	public function test_render_content_should_render_static() {
+		require_once __DIR__ . '/mock/mock-widget.php';
```
Comment (matipojo) [inline tests/phpunit/elementor/includes/base/test-widget-base.php:16]: Autoload?

File: tests/phpunit/elementor/includes/base/test-widget-base.php [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+		$widget = new Mock_Widget( [ 'settings' => [], 'id' => '1' ], [] );
+
+		ob_start();
+
+		$widget->render_content();
+
+		$this->assertRegExp( '/render/', ob_get_clean() );
+	}
+}
```
Comment (matipojo) [inline tests/phpunit/elementor/includes/base/test-widget-base.php:56]: Skin test?

**Review by matipojo — CHANGES_REQUESTED** [review]

File: tests/phpunit/elementor/includes/base/test-widget-base.php (hunk 4 above)
Comment (Nevoss) [reply tests/phpunit/elementor/includes/base/test-widget-base.php:16]: Not working in tests. 
The core files are loaded throw the autoloader.
The tests file loaded only when there is a 'test-' prefix

**Review by matipojo — APPROVED** [review]

**Review by kobizz — CHANGES_REQUESTED** [review]

File: includes/base/widget-base.php [hunk 6]
```diff
@@ -525,9 +525,9 @@ public function render_content() {
 		$skin = $this->get_current_skin();
 		if ( $skin ) {
 			$skin->set_parent( $this );
-			$skin->render();
+			$skin->render_by_mode();
 		} else {
-			$this->render();
+			$this->render_by_mode();
 		}
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply includes/base/widget-base.php:531]: It the first one we call the `render_by_mode` on "skin"
on the second one we call on "this".

[the remark this answers is not in the document]
Comment (kobizz) [reply includes/base/widget-base.php:531]: You are right. My bad...

**Review by kobizz — APPROVED** [review]

**Pull request merged**